tracer/src/Datadog.Trace/Baggage.cs

```diff
@@ -416,6 +418,28 @@ public void MergeInto(Baggage destination)
         }
     }
 
+    public void ForEach<T>(T processor)
```

Comment:
What is this ForEach used for?

Reply:
It's used by the propagator to "iterate" through the items, but it's "reversed" so that the iterating happens internally in `Baggage` instead of externally, if that makes sense. This allow us to use the `List<T>` iterator which doesn't incur heap allocations or boxing, and lets us handle the locking logic internally as well.

It's a pattern (introduced by @tonyredondo) that we use in several places. For example, see the `TagWriter` and `TraceTagWriter` types here:
https://github.com/DataDog/dd-trace-dotnet/blob/d7f025c96f27039d633cfdaed2302dcaee2273a4/tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs#L729-L831

And `TraceTagSetter` here:
https://github.com/DataDog/dd-trace-dotnet/blob/d7f025c96f27039d633cfdaed2302dcaee2273a4/tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs#L318

The examples above use a method like `Process()` or `Next()` to handle each item in the collection, but they are tailored to working with span tags or trace tags. I wanted to go with a more generic interface. At first I tried using the BCL's `IObserver`, but I needed a way to stop the iteration when we hit either of the baggage size limits, so I extended that into `ICancellableObserver`.

I just noticed in other places we use the name `Enumerate()` instead of `ForEach()`. I honestly didn't put much thought into this name, so happy to rename it to `Enumerate()` that if it makes more sense.

Hopefully one day we can go back and refactor most places where we use custom versions of this pattern to be more consistent and use `IObserver` (or `ICancellableObserver`).😅

Reply:
For it's usage, see

`W3CBaggagePropagator.CreateHeader()` here:
https://github.com/DataDog/dd-trace-dotnet/blob/482120a14ac43aec2ee85caa16a36becd510c82a/tracer/src/Datadog.Trace/Propagators/W3CBaggagePropagator.cs#L224

and `W3CBaggageHeaderBuilder` here:
https://github.com/DataDog/dd-trace-dotnet/blob/482120a14ac43aec2ee85caa16a36becd510c82a/tracer/src/Datadog.Trace/Propagators/W3CBaggagePropagator.cs#L286C20-L353

Reply:
Thanks for the explanation! I completely missed the usage in the propagator so that's why I was curious.

And `Enumerate()` would be more helpful for consistency 😅

Reply:
Yes is a pattern to avoid any allocation while iterating over all collections and keep everything stack allocated. I came up with that after several dotteace/dotmemory roundtrips.

This is important because writing tags and metrics is the hottest path in the tracer.

Reply:
Renamed `ForEach()` to `Enumerate()`.

> This is important because writing tags and metrics is the hottest path in the tracer.

Yeah, this is for injecting baggage items into headers, which I don't expect many people to use (so not the hottest path), but any little bit counts.

> I don't know how performant would be IObserver but I doubt it will better than existing code.

Agreed. I just wanted to reuse something that already exists in the BCL instead of coming up with yet another custom interface for this pattern, heh.